File: data-struc-algorithm/sorts/Sorts.js
/**
 * 冒泡、插入、选择排序
 */

// 冒泡排序
const bubbleSort = arr => {
  if (arr.length <= 1) return;
  const n = arr.length;
  for (let i = 0; i < n; i++) {
    let hasChanged = false;

    for (let j = 0; j < n - i - 1; j++) {
      if (arr[j] > arr[j + 1]) {
        // const temp = arr[j];
        // arr[j] = arr[j + 1];
        // arr[j + 1] = temp;
        [arr[j], arr[j + 1]] = [arr[j + 1], arr[j]];
        hasChanged = true;
      }
    }

    if (!hasChanged) {
      break;
    }
  }
  return arr;
  3;
};

// 插入排序
const insertSort = arr => {
  if (arr.length <= 1) return;

  for (let i = 1; i < arr.length; i++) {
    let position = i;
    const value = arr[i];

    while (i >= 0 && arr[position - 1] > value) {
      [arr[position], arr[position - 1]] = [arr[position - 1], arr[position]];
      position--;
    }

    arr[position] = value;
  }
  return arr;
};

// 选择排序
const selectionSort = arr => {
  if (arr.length <= 1) return;

  for (let i = 0; i < arr.length; i++) {
    let minIndex = i;
    for (let j = i + 1; j < arr.length; j++) {
      if (arr[j] < arr[minIndex]) {
        minIndex = j;
      }
    }

    const temp = arr[i];
    arr[i] = arr[minIndex];
    arr[minIndex] = temp;
  }
  return arr;
};

const test = [4, 5, 6, 3, 2, 1];
// console.log(bubbleSort(test));
const testSort = [4, 1, 6, 3, 2, 1];
// console.log(insertSort(testSort));
const testSelect = [4, 8, 6, 3, 2, 1, 0, 12];
console.log(selectionSort(testSelect));

File: data-struc-algorithm/sorts/Sorts.java
/**
 * 冒泡排序、插入排序、选择排序
 */

public class Sorts {

  // 冒泡排序，a 是数组，n 是数组大小
  public static void bubbleSort(int[] a, int n) {
    if (n <= 1)
      return;

    for (int i = 0; i < n; ++i) {
      // 提前退出标志位
      boolean flag = false;

      for (int j = 0; j < n - i - 1; ++j) {
        if (a[j] > a[j + 1]) { // 交换
          int tmp = a[j];
          a[j] = a[j + 1];
          a[j + 1] = temp;
          // 此次冒泡有数据交换
          flag = true;
        }
      }
      // 没有数据交换，直接退出
      if (!flag)
        break;
    }
  }

  // 插入排序
  public static void insertSort(int[] a, int n) {
    if (n <= 1)
      return;

    for (int i = 1; i < n; ++i) {
      int value = a[i];
      int j = i - 1;
      // 查找要插入的位置并移动数据
      for (; j >= 0; --j) {
        if (a[j] > value) {
          a[j + 1] = a[j]; // 数据移动
        } else {
          break;
        }
      }
      a[j + 1] = value; // 插入数据
    }
  }

  // 选择排序
  public static void selectionSort(int[] a, int n) {
    if (n <= 1)
      return;

    for (int i = 0; i < n - 1; i++) {
      int minMax = n;
      for (int j = i + 1; j < n; ++j) {
        if (a[j] < a[minMax]) {
          minMax = j;
        }
      }

      // 交换
      int tmp = a[i];
      a[i] = a[minMax];
      a[minMax] = tmp;

    }
  }
}